### PR TITLE
Facilicate features testing for SEV/SEV-ES guest

### PR DESCRIPTION
**Pull request by waynechen55**
* **Power** on guest if it can not be reached for a while and redo ssh port detecting  if $state_check is given non-zero value explicitly in wait_guest_online in lib/virt_autotest/utils.pm

* **Adjust** timeout and retry values adjustment to some script_retry calls to make sure enough attempts

* **Keep** nvram when undefine uefi guest to ensure the guest can be redefined and then started up again

* **Ensure** guest is in shut-off state if it can not be destroyed and exclude nvram from "virsh pool-list

*  **Skip** SR-IVO/vCPU Hotplugging/Internal Snaphsot/External Snapshot tests for SEV/SEV-ES guests due to lacking of support

* **Verification runs:**
  * [sev-es 15sp3 on 15sp4](http://10.67.131.12/tests/438)
  * [sev-es 15sp4 on 15sp4](http://10.67.131.12/tests/439)
  * [uefi 12sp5 on 15sp4](https://openqa.suse.de/tests/9091409)
  * [uefi 15sp3 on 15sp4](https://openqa.suse.de/tests/9091411)
  * [uefi 15sp4 on 15sp4](https://openqa.suse.de/tests/9091414)
  * [uefi 15sp4 on 12sp5](https://openqa.suse.de/tests/9085532)
  * [uefi 15sp4 on 15sp3](https://openqa.suse.de/tests/9093206)
  * [non-uefi 15sp4 on 15sp3](https://openqa.suse.de/tests/9125705)
  * [non-uefi 15sp4 on 12sp5](https://openqa.suse.de/tests/9125745)
  * [non-uefi 12sp5 on 15sp4](http://10.67.131.12/tests/584)
  * [non-uefi 15sp3 on 15sp4](http://10.67.131.12/tests/587)
  * [non-uefi 15sp4 on 15sp4](http://10.67.131.12/tests/593)

### CODE REVIEW

**Comment by waynechen55** [comment]
@alice-suse @Julie-CAO @guoxuguang @nanzhg Please have a look.

File: lib/virt_autotest/utils.pm
Comment (alice-suse) [inline lib/virt_autotest/utils.pm:434]: @tonyyuan1  @tbaev  This change may impact maintenance test. Would you like to have a check if all is good with this change?

File: lib/virt_autotest/utils.pm
Comment (alice-suse) [inline lib/virt_autotest/utils.pm:460]: Would you please explain why timeout is needed when retry is customized and delay fixed?

**Review by alice-suse — COMMENTED** [review]
Generally LGTM. But there are quite some detailed changes which would potentially impact maintenance test and existing feature test. I guess all relevant automation code owners are interested to give it a review  @tonyyuan1 @tbaev  @Julie-CAO  @guoxuguang?

File: lib/virt_autotest/utils.pm
Comment (waynechen55) [reply lib/virt_autotest/utils.pm:460]: Command had ever failed due to timed out.

File: lib/virt_autotest/virtual_network_utils.pm
Comment (waynechen55) [reply lib/virt_autotest/virtual_network_utils.pm:251]: Most guests will be "virsh undefined" successfully. Failed ones to be tried with "--keep-nvram" to try again.  It is easy way to handle all scenarios.

**Comment by guoxuguang** [comment]
refer to this PR, find out that keep nvram when undefine sev(es) guest. would you mind help us make sure if this PR code change work for existed OSD sle guest without uefi test run. thanks

**Comment by waynechen55** [comment]
> refer to this PR, find out that keep nvram when undefine sev(es) guest. would you mind help us make sure if this PR code change work for existed OSD sle guest without uefi test run. thanks

There is no impact on traditional guests. Firstly, traditional guest will be directly undefined without "--keep-nvram". Secondly, operations on traditional guest will ignore "--keep-nvram". For example, my own traditional guest (non uefi):

virsh undefine opensusetumbleweed-03 --keep-nvram 
Domain 'opensusetumbleweed-03' has been undefined

virsh list --all
 Id   Name                    State
----------------------------------------
 -    opensusetumbleweed-01   shut off
 -    opensusetumbleweed-02   shut off

**Comment by guoxuguang** [comment]
> > refer to this PR, find out that keep nvram when undefine sev(es) guest. would you mind help us make sure if this PR code change work for existed OSD sle guest without uefi test run. thanks
> 
> There is no impact on traditional guests. Firstly, traditional guest will be directly undefined without "--keep-nvram". Secondly, operations on traditional guest will ignore "--keep-nvram". For example, my own traditional guest (non uefi):
> 
> ```
> virsh list --all
>  Id   Name                    State
> ----------------------------------------
>  -    opensusetumbleweed-01   shut off
>  -    opensusetumbleweed-02   shut off
>  -    opensusetumbleweed-03   shut off
> 
> virsh undefine opensusetumbleweed-03 --keep-nvram 
> Domain 'opensusetumbleweed-03' has been undefined
> 
> virsh list --all
>  Id   Name                    State
> ----------------------------------------
>  -    opensusetumbleweed-01   shut off
>  -    opensusetumbleweed-02   shut off
> ```
Well, just clearly confirm that "--keep-nvram" option work for our support guests - sles12+, sles15+ , sles qam and opensuse guests when vitrsh undefine $guest.  Because of, just only use with  "--keep-nvram" option as the workaround to undefine sles uefi guest from my side.  I do not think "--keep-nvram" as the universal option to use for us. 

FYI
       --nvram  and --keep-nvram specify accordingly to delete or keep nvram (/domain/os/nvram/) file. If the domain
       has an nvram file and the flags are omitted, the undefine will fail.

File: lib/virt_autotest/utils.pm
Comment (Julie-CAO) [inline lib/virt_autotest/utils.pm:438]: I concern this makes too big changes to wait_guest_online(), It doubles the waiting time and restarts the guest if first try fails. I would like to suggest to create a new function as wait_guest_online() is wide used in virt test.

In the case of sriov network card test, there is a test step, attach an vf card to guest then reboot the guest. This function is used to check if guest reboots successfully. In [this failed test](https://openqa.nue.suse.com/tests/8751801#step/sriov_network_card_pci_passthrough/1051), it took 30 minutes and lots of screenshots were printed when guest failed to bootup. but with this change, the test would be aborted after 60 minutes. I don't think it is afordable.
Meanwhile, if a device is hotplugged to the guest in live mode(in contrast to persistent/config mode), virsh destroy + virsh start may not bring up the original guest, it depends on hypervisor.

Comment (waynechen55) [reply lib/virt_autotest/utils.pm:438]: Generally speaking, it does not double waiting time. This change is to check whether the guest is in running after failure. If it is not in running, then it will be started forcibly to be given a second chance to retry.

Comment (Julie-CAO) [reply lib/virt_autotest/utils.pm:438]: but I found another issue. For example, my test is to see whether the guest can reboot normally after a series of operations. Your retry may cover up the problem.

File: lib/virt_autotest/utils.pm
Comment (Julie-CAO) [reply lib/virt_autotest/utils.pm:450]: $timeout is for each execution of 'virsh list ...', rather than the total time of the script_retry().

File: lib/virt_autotest/virtual_network_utils.pm
Comment (Julie-CAO) [inline lib/virt_autotest/virtual_network_utils.pm:125]: Do we really need retry 30 times to try to bring up the interface? if you run tests in beijing subnet, dhcp ip run out sometime. In this case waiting for longer time may help, but this is not a normal situation and our test should not consider it.

**Comment by Julie-CAO** [comment]
@alice-suse The PR looks good to me except some minor issues which I had posted inline. Thank you, @waynechen55 .

File: lib/virt_autotest/utils.pm
Comment (waynechen55) [reply lib/virt_autotest/utils.pm:462]: I double checked, this is necessary. This 'timeout' will be used in script_retry as below:
```
script_retry("timeout $timeout $cmd", $timeout + 3, xxxx)
```

File: lib/virt_autotest/utils.pm
Comment (waynechen55) [reply lib/virt_autotest/utils.pm:440]: I added $state_check which is 0 by default:
```
        # Ensure guest is running
        if (($state_check != 0) and (script_run("virsh list --name --state-running | grep $guest") != 0)) {
            script_run("virsh destroy $guest");
            assert_script_run("virsh start $guest");
            script_retry("nmap $guest -PN -p ssh | grep open", delay => 1, retry => $retries);
        }
```

Comment (Julie-CAO) [reply lib/virt_autotest/utils.pm:440]: then that's ok for me.

**Comment by waynechen55** [comment]
> > > > refer to this PR, find out that keep nvram when undefine sev(es) guest. would you mind help us make sure if this PR code change work for existed OSD sle guest without uefi test run. thanks
> > > 
> > > 
> > > There is no impact on traditional guests. Firstly, traditional guest will be directly undefined without "--keep-nvram". Secondly, operations on traditional guest will ignore "--keep-nvram". For example, my own traditional guest (non uefi):
> > > ```
> > > virsh list --all
> > >  Id   Name                    State
> > > ----------------------------------------
> > >  -    opensusetumbleweed-01   shut off
> > >  -    opensusetumbleweed-02   shut off
> > >  -    opensusetumbleweed-03   shut off
> > > 
> > > virsh undefine opensusetumbleweed-03 --keep-nvram 
> > > Domain 'opensusetumbleweed-03' has been undefined
> > > 
> > > virsh list --all
> > >  Id   Name                    State
> > > ----------------------------------------
> > >  -    opensusetumbleweed-01   shut off
> > >  -    opensusetumbleweed-02   shut off
> > > ```
> > 
> > 
> > Well, just need to clearly confirm that "--keep-nvram" option work for our support guests - sles12+, sles15+ , sles qam and opensuse guests when vitrsh undefine $guest. Because of, just only use with "--keep-nvram" option as the workaround to undefine sles uefi guest from my side. I do not think "--keep-nvram" as the universal option to use for us.
> > FYI --nvram and --keep-nvram specify accordingly to delete or keep nvram (/domain/os/nvram/) file. If the domain has an nvram file and the flags are omitted, the undefine will fail.
> 
> As I said, this usage is safe. I am also running some verification runs on traditional guests to make it more clear.

All non-uefi verification look good:
[non-uefi 15sp4 on 15sp3](https://openqa.suse.de/tests/9125705)
[non-uefi 15sp4 on 12sp5](https://openqa.suse.de/tests/9125745)
[non-uefi 12sp5 on 15sp4](http://10.67.131.12/tests/584)
[non-uefi 15sp3 on 15sp4](http://10.67.131.12/tests/587) 
[non-uefi 15sp4 on 15sp4](http://10.67.131.12/tests/593)

**Comment by alice-suse** [comment]
@Julie-CAO @guoxuguang  @tonyyuan1 @tbaev Hello guys, would you please double confirm if the PR is ready to be merged from your POV? Please raise comments for needed changes before EOB today if you have any, otherwise I will give it the merge ;).

**Comment by Julie-CAO** [comment]
> @Julie-CAO @guoxuguang @tonyyuan1 @tbaev Hello guys, would you please double confirm if the PR is ready to be merged from your POV? Please raise comments for needed changes before EOB today if you have any, otherwise I will give it the merge ;).

The PR LGTM

**Comment by tbaev** [comment]
I don't see anything wrong with the PR, but I think it will be a good idea to make a VR run for KVM and XEN against our tests before merge.

**Comment by waynechen55** [comment]
> I don't see anything wrong with the PR, but I think it will be a good idea to make a VR run for KVM and XEN against our tests before merge.

There are already VRs for kvm and xen involving 12sp5/15sp3/15sp4 host/guest matrix:
  * [uefi 12sp5 on 15sp4 kvm](https://openqa.suse.de/tests/9091409)
  * [uefi 15sp3 on 15sp4 kvm](https://openqa.suse.de/tests/9091411)
  * [uefi 15sp4 on 15sp4 kvm](https://openqa.suse.de/tests/9091414)
  * [uefi 15sp4 on 12sp5 xen](https://openqa.suse.de/tests/9085532)
  * [uefi 15sp4 on 15sp3 kvm](https://openqa.suse.de/tests/9093206)
  * [non-uefi 15sp4 on 15sp3 kvm](https://openqa.suse.de/tests/9125705)
  * [non-uefi 15sp4 on 12sp5 xen](https://openqa.suse.de/tests/9125745)
  * [non-uefi 12sp5 on 15sp4 kvm](http://10.67.131.12/tests/584)
  * [non-uefi 15sp3 on 15sp4 kvm](http://10.67.131.12/tests/587)
  * [non-uefi 15sp4 on 15sp4 xen](http://10.67.131.12/tests/593)

**Comment by guoxuguang** [comment]
> @Julie-CAO @guoxuguang @tonyyuan1 @tbaev Hello guys, would you please double confirm if the PR is ready to be merged from your POV? Please raise comments for needed changes before EOB today if you have any, otherwise I will give it the merge ;).

LGTM

**Review by tbaev — APPROVED** [review]

**Comment by alice-suse** [comment]
Big thanks to all, merge it.

**Pull request merged**
